core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -646,16 +677,37 @@ private[chisel3] object Builder extends LazyLogging {
     }
   }
 
+  // Builds a RenameMap for all Views that do not correspond to a single Data
+  // These Data give a fake ReferenceTarget for .toTarget and .toReferenceTarget that the returned
+  // RenameMap can split into the constituent parts
+  private[chisel3] def makeViewRenameMap: RenameMap = {
+    val renames = RenameMap()
+    for (view <- unnamedViews) {
+      val localTarget = view.toTarget
+      val absTarget = view.toAbsoluteTarget
+      val elts = getRecursiveFields.lazily(view, "")
+        .collect { case (elt: Element, _) => elt }
+      for (elt <- elts) {
+        val targetOfView = reify(elt)
+        renames.record(localTarget, targetOfView.toTarget)
+        renames.record(absTarget, targetOfView.toAbsoluteTarget)
+      }
+    }
+    renames
+  }
+
   private [chisel3] def build[T <: RawModule](f: => T, dynamicContext: DynamicContext): (Circuit, T) = {
     dynamicContextVar.withValue(Some(dynamicContext)) {
+      ViewParent // Must initialize the singleton in a Builder context or weird things can happen
+                 // in tiny designs/testcases that never access anything in chisel3.internal
       checkScalaVersion()
       logger.warn("Elaborating design...")
       val mod = f
       mod.forceName(None, mod.name, globalNamespace)
       errors.checkpoint(logger)
       logger.warn("Done elaborating.")
 
-      (Circuit(components.last.name, components.toSeq, annotations.toSeq), mod)
+      (Circuit(components.last.name, components.toSeq, annotations.toSeq, makeViewRenameMap), mod)
```

Comment:
Now that we are returning a renamemap, perhaps we can fix up our naming stuff so that you can do .toTarget during elaboration, and eliminate ChiselAnnotation?

Reply:
In a future PR, obviously.

Reply:
While we could do that under your old proposal, I would say we can make `.toTarget` work during elaboration with the following 2 API changes:
1. Require the compiler plugin -- we've already decided we should do this
2. Do not have ports always win in naming over non-ports

(2) is a little iffier but I think that the compiler plugin and the prefixing in particular make this much more palatable. The fear is a `.suggestName` on a port conflicting with the name of something like a `Wire` that's already had `.toTarget` called on it.